lib/segment/src/index/struct_payload_index.rs

```diff
@@ -308,6 +308,30 @@ impl StructPayloadIndex {
             Condition::Field(field_condition) => self
                 .estimate_field_condition(field_condition, nested_path)
                 .unwrap_or_else(|| CardinalityEstimation::unknown(self.available_point_count())),
+
+            Condition::Abstract(cond) => {
+                // TODO(resharding): I have no idea what I'm doing... 💀
+
+                let id_tracker = self.id_tracker.borrow();
+
+                let ids: HashSet<_> = id_tracker
+                    .iter_internal()
+                    .filter(|&internal_id| {
+                        id_tracker
+                            .external_id(internal_id)
+                            .map_or(false, |external_id| cond(external_id))
+                    })
+                    .collect();
+
+                let num_ids = ids.len();
+
+                CardinalityEstimation {
+                    primary_clauses: vec![PrimaryCondition::Ids(ids)],
+                    min: num_ids,
+                    exp: num_ids,
+                    max: num_ids,
+                }
```

Comment:
I kinda copied it from `HasId` branch above, but now that I think about it, this should be a reasonable estimation? (I'd have to refactor a bit so that I can access `hash_ring.len()`, but it's trivial. Just introduce a custom trait.)

```rust
{
	min: 0, 
	max: id_tracker.external_id.len(),
	exp: id_tracker.external_id.len() * (hash_ring.len() - 1) / hash_ring.len(),
}
```

Not sure about `primary_clauses`, though.

Reply:
I'd believe so.

Though, this estimation is specific to using a hashring. It does not work for a general purpose 'abstract' filter.

Maybe we can rename it to specialize it for filtering by hashring, while keeping the abstract closure.

Reply:
Though, I'd want @generall's opinion on this.

Initially, I've copied `Condition::Abstract`/`Condition::Resharding` implementation from `Condition::HasId` implementation. I've tracked how `primary_clauses` are used, but it's still a bit over my head. I can't decisively reason if initial (e.g., similar to `HasId`) or current implementation would be better for `Condition::Resharding`.

I can only note that during last stages of resharding, `Condition::Resharding` would be applied to **all** read operations. And it may have to be applied for a while, until we delete "resharded" points from the old shards. Or, at best, until we delete "resharded" points *on a particular node*.